Issue#49 Implement BPM Control Class

KevinZhiyuanDing opened the pull request:
I have made the BPM control class. 
The constructor takes the pin number for the BPM control output pin. 
I have added the pin number definition in ```settings.h```:

```
//BPM
#define BPM_CTRL_OUT        3
```

The pi pico outputs a PWM signal from the BPM_CTRL_OUT pin. The signal is augmented by the amplifier and doubled in amplitude, and then send to the BPM control module to control breaking power.

The BPM control class if included below. The pwm has a range of (0,1023) and analog frequency of  20kHz for now (i'm not sure if this frequency is suitable or not):
```
#include "HardwareInput.h"
#include "BpmControl.h"

#define ANALOG_MAX 1023
#define ANALOG_FREQ 20000

BpmControl::BpmControl (pin_size_t pinNum){
    _pinNum = pinNum;
}

void BpmControl::init (){
    pinMode(_pinNum, OUTPUT);
    analogWriteRange(ANALOG_MAX);
    analogWriteFreq(ANALOG_FREQ);
};

void BpmControl::setControlSignal (int32_t controlSignal){ 
    if (_active == true){
        if (controlSignal >= 0 && controlSignal <= 1023){
            analogWrite (_pinNum, controlSignal);
        }
    }
    else {
        analogWrite (_pinNum, 0);
    }
};

void BpmControl::setActive (bool active){
    _active = active;
    if (active == false) {
        analogWrite (_pinNum, 0);
    }
};
```
Its corresponding header file is:
```
#ifndef _BPM_CONTROL_H_
#define _BPM_CONTROL_H_
#include "settings.h"

class BpmControl {
    private:    
        int _pinNum;
        bool _active = false;
    public:
        BpmControl (pin_size_t pinNum);
        void init ();
        void setControlSignal (int32_t signalStrength);
        void setActive (bool state);
};

#endif
```

KevinZhiyuanDing committed:
Issue#49 Implement BPM Control class

KevinZhiyuanDing committed:
Add header file

JamesZhZhang commented:
@KevinZhiyuanDing were you guys able to figure out why build-firmware failed?

cosparks commented:
This must have branched off from dev-v2 around the time [this commit](https://github.com/supermileage/dynamometer-firmware/commit/faf0ae63f6da219ab5ec03a4b04ed0e1a41ee0c2) had been made.

This line is still uncommented in settings.h
```c++
#define TOUCH_CS 14
```

which defines a macro used by the TFT_eSPI library. The error msg tells you why the build failed:
`#error Touch screen not supported in parallel or SPI PIO mode, use a separate library.`

The pico is using PIO SPI to communicate with the display, so touch screen functionality isn't available. I saw that this has been fixed on dev-v2. If you merge latest changes from dev-v2 onto this branch, you will probably see that line show up in the diff.

KevinZhiyuanDing committed:
Update settings.h

cosparks commented on src/settings.h:
```diff
@@ -28,7 +28,7 @@
 #define LCD_DC      7
 #define LCD_CS      5
-#define TOUCH_CS    14
+//#define TOUCH_CS    14
```
I think you have a merge conflict here, so you'll need to merge dev-v2 into this branch

cosparks commented on src/System/BpmControl.cpp:
```diff
@@ -0,0 +1,33 @@
+    if (_active == true){
+        if (controlSignal >= 0 && controlSignal <= 1023){
+            analogWrite (_pinNum, controlSignal);
+        }
```
If the control signal is > 1023 or < 0, then the calling code isn't using this API properly.  You should probably print an error message here or make an assertion instead of silently ignoring bad input values. It might help you catch a bug or two in the future.

The [error logger header](https://github.com/supermileage/dynamometer-firmware/blob/dev/src/System/ErrorLogger.h) has a number of macros which you can use

cosparks commented on src/System/BpmControl.cpp:
```diff
@@ -0,0 +1,33 @@
+    analogWriteFreq(ANALOG_FREQ);
+};
+
+void BpmControl::setControlSignal (int32_t controlSignal){ 
```
The spacing before the parens here and in the header doesn't follow the style of the rest of the supermileage codebase. The style guide is [here](https://www.notion.so/C-Style-Guide-f40ef59a0c3140148e1e578bfc61f1b8).

It might be good to set up a linter eventually.

cosparks commented:
I just had a couple comments on the code, but other than that this looks good.

Have you tested this on the new board to make sure that you're actually getting 0-10v on the BPM control pin?  I would maybe do that before submitting this code.

KevinZhiyuanDing committed:
Fix merge conflict, include assertion to control signal input, and change code style to fit standard

KevinZhiyuanDing committed:
Update HardwareRotaryEncoder.h

Quick fix to change the counter to rotary encoder from unsigned integer to signed integer.

valentino-jaber committed:
Merge branch 'dev-v2' into BPMcontrol

valentino-jaber reviewed (approved):
Looks good. Just resolved a merge conflict

cosparks commented on src/System/BpmControl.cpp:
```diff
@@ -0,0 +1,37 @@
+            analogWrite(_pinNum, controlSignal);
+        }
+        else {
+            dyno_assert(controlSignal >= 0 && controlSignal <= 1023);
```
asserts shouldn't be placed in an if/else block. the whole idea is that you can compile them out if you want to (when building for production, for example). if you do that in this case then you get an empty else condition.

I would just place dyno_assert before the call to analogWrite

cosparks replied:
I recommend:

```c++
// ...
    if (_active = true){
        dyno_assert(controlSignal >= 0 && controlSignal <= 1023);
        analogWrite(_pinNum, controlSignal);
    } else {
    // ...
```

cosparks commented:
Other than the placement of the assert this all looks good

KevinZhiyuanDing committed:
Update BpmControl.cpp Asserts

Fix the placement of the assert statement used to test if control signal of BPM is within a range that makes sense.
